parsers/CR.py

```diff
@@ -179,85 +77,84 @@ def fetch_production(
     r = Session()
-    url = "https://apps.grupoice.com/CenceWeb/CencePosdespachoNacional.jsf"
-    response = r.get(url)
-
-    soup = BeautifulSoup(response.text, "html.parser")
-    jsf_view_state = soup.find("input", {"name": "javax.faces.ViewState"})["value"]
+    day = arw_datetime.format("DD")
+    month = arw_datetime.format("MM")
+    year = arw_datetime.format("YYYY")
+    url = f"https://apps.grupoice.com/CenceWeb/data/sen/json/EnergiaHorariaFuentePlanta?anno={year}&mes={month}&dia={day}"
 
-    data = [
-        ("formPosdespacho:txtFechaInicio_input", target_datetime.format(DATE_FORMAT)),
-        ("formPosdespacho:pickFecha", ""),
-        ("formPosdespacho_SUBMIT", 1),
-        ("javax.faces.ViewState", jsf_view_state),
-    ]
-    response = r.post(url, data=data)
+    response = r.get(url)
+    resp_data = response.json()["data"]
 
-    # tell pandas which table to use by providing CHARACTERISTIC_NAME
-    df = pd.read_html(
-        response.text, match=CHARACTERISTIC_NAME, skiprows=1, index_col=0
-    )[0]
+    results = {}
+    for hourly_item in resp_data:
+        if hourly_item["fuente"] == "Intercambio":
+            continue
+        if hourly_item["fecha"] not in results:
+            results[hourly_item["fecha"]] = empty_record(zone_key)
 
-    results = df_to_data(zone_key, target_datetime, df, logger)
+        results[hourly_item["fecha"]]["datetime"] = arrow.get(
+            hourly_item["fecha"], tzinfo=TIMEZONE
+        ).datetime
+        results[hourly_item["fecha"]]["production"][
+            SPANISH_TO_ENGLISH[hourly_item["fuente"]]
+        ] += hourly_item["dato"]
 
-    return results
+    return [results[k] for k in sorted(results.keys())]
 
 
 def fetch_exchange(
     zone_key1: str = "CR",
-    zone_key2: str = "NI",
+    zone_key2: str = "PA",
     session: Optional[Session] = None,
     target_datetime: Optional[datetime] = None,
     logger: Logger = getLogger(__name__),
 ) -> dict:
-    """Requests the last known power exchange (in MW) between two regions."""
+    """Gets an exchange pair from the SIEPAC system."""
     if target_datetime:
         raise NotImplementedError("This parser is not yet able to parse past dates")
 
-    sorted_zone_keys = "->".join(sorted([zone_key1, zone_key2]))
+    sorted_zones = "->".join(sorted([zone_key1, zone_key2]))
+
+    if sorted_zones not in EXCHANGE_JSON_MAPPING.keys():
+        raise NotImplementedError("This exchange is not implemented.")
 
-    df = pd.read_csv(
-        "http://www.enteoperador.org/newsite/flash/data.csv", index_col=False
-    )
+    s = session or Session()
 
-    if sorted_zone_keys == "CR->NI":
-        flow = df["NICR"][0]
-    elif sorted_zone_keys == "CR->PA":
-        flow = -1 * df["CRPA"][0]
-    else:
-        raise NotImplementedError("This exchange pair is not implemented")
+    raw_data = s.get(EXCHANGE_URL).json()
+    flow = round(extract_exchange(raw_data, sorted_zones), 1)
```

Comment:
Since `extract_exchange` can return None we can't call round on it directly. We need to make sure it's a number first.